### PR TITLE
Data-Parallel Deepseek Model

### PR DESCRIPTION
**Pull request by barci2**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/28918)

### What's changed
- Moved the modules around and renamed them  
- Created a DP version of MLA, Embedding, (MoE/MLP) Decoder Block and Model
- Added the new classes to tests

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes (unrelated failures)
- [ ] [Deepseek](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-deepseek-tests.yaml) CI passes

### CODE REVIEW

**Review by kpaigwar — APPROVED** [review]
Nice 🔥

File: models/demos/deepseek_v3/tt/generator.py [hunk 1]
```diff
@@ -249,25 +250,21 @@ def _decode_step(self, tokens_step: torch.Tensor, positions: torch.Tensor) -> to
         logits = ttnn.to_torch(logits_tt, mesh_composer=ttnn.ConcatMeshToTensor(self.mesh_device, dim=3))
 
         # Free device tensors for this step
-        ttnn.deallocate(tt_tokens)
-        ttnn.deallocate(logits_tt)
-
-        return logits  # [1, 1, B, V]
```
Comment (yieldthought) [inline models/demos/deepseek_v3/tt/generator.py:295]: This looks like a mistake, I assume we actually do want to return the logits!

File: models/demos/deepseek_v3/tt/rope.py [hunk 2]
```diff
@@ -71,31 +71,30 @@ class RotarySetup:
     Class to set up rotary positional embeddings for DeepSeekV3 models.
 
     Duplicate + changes from TTT rope.py
+    NOTE: it would be better to re-migrate this class to TTT rope.py at some point
     """
 
-    def __init__(self, device, batch_size: int, hf_config, use_dp: bool = False):
-        self.batch_size = batch_size
+    def __init__(self, device, batch_size_per_row: int, hf_config):
+        self.batch_size_per_row = batch_size_per_row
```
Comment (yieldthought) [inline models/demos/deepseek_v3/tt/rope.py:78]: [tests/tt_eager/python_api_testing/unit_testing/misc/test_deepseek_mla_ops.py](https://github.com/tenstorrent/tt-metal/pull/30016/files#diff-2feeea142c87f7df452928e2b027010fb9a46ce5ffe9ca2e41751df3a8471e54R540) still uses batch_size= and will fail. There might be other places that still use that too?

File: models/demos/deepseek_v3/tt/rope.py [hunk 3]
```diff
@@ -190,8 +191,8 @@ def get_rot_mats_table(self, seq_len=None):
         )
 
         if seq_len is not None:
-            return cos_matrix, sin_matrix, self.transformation_mat_prefill
-        return cos_matrix, sin_matrix
+            return {"cos_matrix": cos_matrix, "sin_matrix": sin_matrix, "trans_matrix": self.transformation_mat_prefill}
+        return {"cos_matrix": cos_matrix, "sin_matrix": sin_matrix}
```
Comment (yieldthought) [inline models/demos/deepseek_v3/tt/rope.py:195]: test_deepseek_mla_ops.py needs to be updated to expect a dictionary and not a tuple here as well, maybe other consumers of RopeSetup as well?

File: models/demos/deepseek_v3/utils/test_utils.py
Comment (yieldthought) [inline models/demos/deepseek_v3/utils/test_utils.py]: dp_factor has been replaced with mesh_shape so this will fail. Any other calls missed?

File: models/demos/deepseek_v3/utils/config_helpers.py [hunk 4]
```diff
@@ -19,7 +19,7 @@
 
 # Constants
 NORM_CATEGORIES = {"attention_norm", "mlp_norm", "q_norm", "k_norm"}
-MAX_BATCH_SIZE = 32
```
Comment (yieldthought) [inline models/demos/deepseek_v3/utils/config_helpers.py:22]: Basically all the unit tests not touched in this PR still look for MAX_BATCH_SIZE, e.g. models/demos/deepseek_v3/tests/test_mlp_1d_ops.py, test_mlp_1d_dequant_ops.py, test_mlp_non_expert_ops.py, test_mlp_shared_expert_ops.py, test_distributed_rms_norm_ops.py

**Review by yieldthought — CHANGES_REQUESTED** [review]

File: models/demos/deepseek_v3/tt/rope.py (hunk 2 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/rope.py:78]: Nope, that's the only one

File: models/demos/deepseek_v3/utils/config_helpers.py (hunk 4 above)
Comment (barci2) [reply models/demos/deepseek_v3/utils/config_helpers.py:22]: I don't get what you mean. None of the files you mentioned exist.

Comment (yieldthought) [reply models/demos/deepseek_v3/utils/config_helpers.py:22]: My bad, these are auto-generated ops tests in my local working directory from something I was exploring.

File: models/demos/deepseek_v3/tt/generator.py [hunk 5]
```diff
@@ -324,15 +329,10 @@ def _prefill(self, tokens_batched: torch.Tensor) -> torch.Tensor:
             batch_size=1,
```
Comment (yieldthought) [inline models/demos/deepseek_v3/tt/generator.py]: This has been changed to `batch_size_per_row` in RotarySetup. Is this function not covered by any tests?

**Review by yieldthought — CHANGES_REQUESTED** [review]
Also git amend with a proper commit message please, "Implementation" is not ok to see sprawled over so many files in the repo.

File: models/demos/deepseek_v3/tt/generator.py (hunk 5 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/generator.py]: None that run in CI

**Review by yieldthought — APPROVED** [review]
LGTM if the tests go green!

**Review by pprajapatiTT — CHANGES_REQUESTED** [review]
Is demo working with this change? We dont have demo tests in CI. So for now need to test manually.

**Comment by barci2** [comment]
@pprajapatiTT the demo test doesn't work on main either

**Review by pprajapatiTT — APPROVED** [review]

**Comment by pprajapatiTT** [comment]
@barci2 , I am not sure what broke the demo. I will check tomorrow. 

However, first you mentioned demo worked and then updated comment demo is broken on main. Seems like with your branch it was working first but later with latest main something broke it.
I approved the PR, i'll check demo tomorrow.

Feel free to merge PR once CI passes.

**Review by blozano-tt — APPROVED** [review]

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -16,6 +16,7 @@ Summarize the changes made and its impact.
 - [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
 - [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
 - [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
+- [ ] [Galaxy test, for DeepSeek](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-deepseek-tests.yaml) CI passes, if applicable, because of current DeepSeek work
```
Comment (tt-rkim) [inline .github/pull_request_template.md]: We care about this at that level?

Are there small and very fast Galaxy APC tests we can add so we don't break this pipeline and cause issues all the time?

**Comment by kpaigwar** [comment]
What's left to get this in?
